### PR TITLE
Aura block reward contract transitions

### PR DESCRIPTION
**Pull request by LukaszRozmej**
#1407 Add multiple reward contracts based on blocknumbers

### CODE REVIEW

File: src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs [hunk 1]
```diff
@@ -83,6 +90,18 @@ public void constructor_throws_ArgumentNullException_on_null_transactionProcesso
             Action action = () => new AuRaRewardCalculator(_auraParameters, _abiEncoder, null);
             action.Should().Throw<ArgumentNullException>();
         }
+
+        [Test]
+        public void constructor_throws_ArgumentException_on_BlockRewardContractTransition_higher_than_BlockRewardContractTransitions()
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: I did not understand the mame of the test.

File: src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs [hunk 2]
```diff
@@ -83,6 +90,18 @@ public void constructor_throws_ArgumentNullException_on_null_transactionProcesso
...
+        {
+            _auraParameters.BlockRewardContractTransitions = new Dictionary<long, Address>()
+            {
+                {2, Address.FromNumber(2)}
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: does the address 2 mean something here? if not then worth using TestItem.AddressA

Comment (tkstanczak) [reply src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: Address 2 is in the precompiles space

File: src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs [hunk 3]
```diff
@@ -101,7 +120,25 @@ public void calculates_rewards_correctly_after_contract_transition(long blockNum
...
+        public void calculates_rewards_correctly_after_subsequent_contract_transitions(long blockNumber, long expectedReward)
+        {
+            _auraParameters.BlockRewardContractTransitions = new Dictionary<long, Address>()
+            {
+                {50, _address50},
+                {150, _address150}
+            };
+            _block.Header.Number = blockNumber;
+            var expected = new BlockReward(_block.Beneficiary, expectedReward, BlockRewardType.Block);
+            SetupBlockRewards(new Dictionary<Address, BlockReward[]>() {{Address.FromNumber((UInt256) blockNumber), new[] {expected}}});
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: is the address somehow related to block numbees or just testing convenience?

File: src/Nethermind/Nethermind.AuRa/Rewards/AuRaRewardCalculator.cs [hunk 4]
```diff
@@ -80,9 +124,9 @@ private BlockReward[] CalculateRewardsWithContract(Block block)
             }
 
             var (beneficiaries, kinds) = GetBeneficiaries();
-            var transaction = _contract.Reward(beneficiaries, kinds);
-            _contract.InvokeTransaction(block.Header, _transactionProcessor, transaction, _tracer);
-            var (addresses, rewards) = _contract.DecodeRewards(_tracer.ReturnValue);
+            var transaction = contract.Reward(beneficiaries, kinds);
+            contract.InvokeTransaction(block.Header, _transactionProcessor, transaction, _tracer);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Rewards/AuRaRewardCalculator.cs]: can we tename InvokeTransaction to Call or similar so it is obvious that we are not changing state?

File: src/Nethermind/Nethermind.Core/Extensions/IListExtensions.cs [hunk 5]
```diff
@@ -14,12 +14,78 @@
...
+        /// Performs a binary search on the specified collection.
+        /// </summary>
+        /// <typeparam name="TItem">The type of the item.</typeparam>
+        /// <typeparam name="TSearch">The type of the searched item.</typeparam>
+        /// <param name="list">The list to be searched.</param>
+        /// <param name="value">The value to search for.</param>
+        /// <param name="comparer">The comparer that is used to compare the value
+        /// with the list items.</param>
+        /// <returns></returns>
+        public static int BinarySearch<TItem, TSearch>(this IList<TItem> list, TSearch value, Func<TSearch, TItem, int> comparer)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Core/Extensions/IListExtensions.cs]: https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.binarysearch?view=netframework-4.8

File: src/Nethermind/Nethermind.Core/Extensions/IListExtensions.cs [hunk 6]
```diff
@@ -14,12 +14,78 @@
...
 //  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
 
+using System;
 using System.Collections.Generic;
 
 namespace Nethermind.Core.Extensions
 {
     public static class IListExtensions
     {
         public static T GetItemRoundRobin<T>(this IList<T> array, long index) => array.Count == 0 ? default : array[(int) (index % array.Count)];
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Core/Extensions/IListExtensions.cs]: I suggest only moving things to Nethermind.Core when they are used in at elast 2 projects or cause some dependencues issues -I keep pruning the classes that grow here

File: src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: there is an old BlockRewardContractTransition and there is new BlockRewardContractTransitions, they are concatenated, but the singular one (BlockRewardContractTransition) has to be lower than first of the plural one (BlockRewardContractTransitions)

Comment (tkstanczak) [reply src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: the naming is confusing but it is a non-blocker

File: src/Nethermind/Nethermind.Core/Extensions/IListExtensions.cs (hunk 5 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/Extensions/IListExtensions.cs]: Yeah but its List/Array only not IList and lacks delegate support which is more convenient here

**Comment by tkstanczak** [comment]
feel free to merge

**Pull request merged**